// generator-jaywing-atomic/generators/components/lib/helpers/docsDataHelper.js
//-------------------------------------------------------
//  Library: Data Helper
//-------------------------------------------------------

//-------------------------
//  Private
//-------------------------

//-------------------------
//  Public
//-------------------------

var self = module.exports = {

  getDocsDataPageItemId: function (name) {
    var result = name.replace(/\s+/g, '-').toLowerCase();
    return result;
  },

  extractPageObj: function (docsDataObj, categoryName, subCategoryName) {

    var page = null;

    var pages = docsDataObj.docs_pages;
    for (var i = 0; i < pages.length; i++) {
      var category = pages[i].category;
      var section = pages[i].section;
      if(category == subCategoryName && section == categoryName.toLowerCase()){
        page = pages[i];
      }
    }

    return page;

  },

  createPageObj: function (categoryName, subCategoryName) {

    var obj = {};

    obj.category = subCategoryName;
    obj.section = categoryName.toLowerCase();
    obj.items = [];

    return obj;

  },

  createPageItemObj: function (name, categoryName) {

    var obj = {};

    obj.id = this.getDocsDataPageItemId(name);
    obj.title = name;
    obj.section = categoryName.toLowerCase();
    obj.ready = true;

    return obj;

  },

  addPageObjToDocsDataObj: function (docsDataObj, pageObj) {

    docsDataObj.docs_pages.push(pageObj);

  },

  addPageItemObjToPageObj: function (pageObj, pageItemObj) {

    pageObj.items.push(pageItemObj);

  }

};
