// generator-jaywing-atomic/lib/components/Core.js
//-------------------------------------------------------
//  Library: Data Manager
//-------------------------------------------------------

var _config = require('../../config');
var _DataHelper = require('../../lib/components/DataHelper');
var _DocsDataHelper = require('../../lib/components/DocsDataHelper');
var _FileMiner = require('../../lib/components/FileMiner');
var _FileManager = require('../../lib/components/FileManager');

//-------------------------
//  Private
//-------------------------

var _data = null;

//-------------------------
//  Public
//-------------------------

var self = module.exports = {

  init: function (generator) {

    _data = {};
    _FileManager.init(generator);

  },

  getCategoryNames: function () {
    var names = _DataHelper.extractDataCategoryNames(_config.components.categories);
    return names;
  },

  getSubCategoryNames: function (categoryName) {

    var names = [];

    // Find the corresponding category in the data
    if(categoryName == 'Core'){
      var categories = _config.components.categories;
      for (var i = 0; i < categories.length; i++) {
        if(categories[i].name == categoryName) {
          var category = categories[i];
          names = _DataHelper.extractDataSubCategoryNames(category);
          break;
        }
      }
    } else {
      names = _FileMiner.getSubCategoryNamesFromSassFolders(categoryName);
    }

    return names;

  },

  getNames: function (categoryName, subCategoryName) {
    var names = _FileMiner.getNamesFromSassFolders(categoryName, subCategoryName);
    return names;
  },

  getSizeNames: function () {
    var names = _DataHelper.extractDataSizeNames(_config.components.sizes);
    return names;
  },

  getModuleClassNames: function () {
    var names = _FileMiner.getModuleClassNamesFromModulesFolder();
    return names;
  },

  add: function (categoryName, subCategoryName, name, description, sizeNames, moduleClassName) {

    var inputDataObj = _DataHelper.createDataObj(categoryName, subCategoryName, name, description, sizeNames, moduleClassName);

    //_FileManager.writeDocsData(inputDataObj);
    _FileManager.addSassFiles(inputDataObj);



    // Add base files
    /*
    for(var i = 0; i < fileExtensions.length; i++) {
      var extension = fileExtensions[i];
      if(extension !== '.scss') {

        var filePath;
        switch(extension) {
          case '.njk':
            filePath = _pathTemplate + _namePath;
            break;
          case '.definition.scss':
            filePath = _pathSass + _categoryPath + '/' + _namePath + '/' + _filePrefix + '.' + _namePath;
            break;
          case '.size.scss':
            filePath = _pathSass + _categoryPath + '/' + _namePath + '/' + _filePrefix + '.' + _namePath;
            break;
          case '.js':
            filePath = _pathJs + _jsClassName;
            break;
        }

        var templatePath = _templateFilePrefix + extension;
        var fullFilePath;

        if(extension == '.definition.scss') sassImports.push(_filePrefix + '.' + _namePath + '.definition');
        if(extension == '.size.scss') {
          for (var a = 0; a < _sizes.length; a++) {
            data.sizeName = _sizeNames[a];
            data.sizeIncludeName = _sizeIncludeNames[a];
            data.sizeCssClassName = _sizeCssClassNames[a];
            var sizePath = _sizes[a].toLowerCase();
            fullFilePath = filePath + '.' + sizePath + '.scss';
            sassImports.push(_filePrefix + '.' + _namePath + '.' + sizePath);
            this.fs.copyTpl(this.templatePath(templatePath), this.destinationPath(fullFilePath), data);
          }
        } else {
          fullFilePath = filePath + extension;
          this.fs.copyTpl(this.templatePath(templatePath), this.destinationPath(fullFilePath), data);
        }
      }
    }

    // Add sass config
    var extension = '.scss';
    var templatePath = _templateFilePrefix + extension;
    var filePath = _pathSass + _categoryPath + '/' + _namePath + '/' + _filePrefix + '.' + _namePath;
    var fullFilePath = filePath + extension;
    data.sassImports = '';
    for (var x = 0; x < sassImports.length; x++) {
      data.sassImports += '@import \'' + sassImports[x] + '\';\n';
    }
    this.fs.copyTpl(this.templatePath(templatePath), this.destinationPath(fullFilePath), data);

    this.log('  Adding ' + _name + ' component into the ' + _category + ' category...');
    */

  },

  delete: function (categoryName, name) {

  }

};
