// generator-jaywing-atomic/lib/components/DataHelper.js
//-------------------------------------------------------
//  Library: Data Helper
//-------------------------------------------------------

var _ = require('lodash');

var _config = require('../../config');
var _DocsDataHelper = require('../../lib/components/DocsDataHelper');
var _PropertyHelper = require('../../lib/components/PropertyHelper');

//-------------------------
//  Private
//-------------------------

function _createSizeObjs(name, sizeNames, sassPath) {
  var objs = [];
  for (var i = 0; i < sizeNames.length; i++) {
    var obj = {};
    obj.componentName = name;
    obj.name = sizeNames[i];
    obj.cssClassName = _PropertyHelper.getSizeCssClassName(name, sizeNames[i]);
    obj.sassIncludName = _PropertyHelper.getSizeSassIncludeName(sizeNames[i]);
    obj.sassFileName = _PropertyHelper.getSassSizeFileName(name, sizeNames[i]);
    obj.sassFilePath = sassPath + obj.sassFileName;
    objs.push(obj);
  }
  return objs;
}

//-------------------------
//  Public
//-------------------------

var self = module.exports = {

  extractDataCategoryNames: function (categoriesData) {
    var names = [];
    for(var i = 0; i < categoriesData.length; i++) {
      names.push(categoriesData[i].name);
    }
    return names;
  },

  extractDataSubCategoryNames: function (categoryData) {
    var names = [];
    var subCategories = categoryData.categories;
    for (var i = 0; i < subCategories.length; i++) {
      var subCategory = subCategories[i];
      var name = subCategory.name;
      names.push(name);
    }
    return names;
  },

  extractDataSizeNames: function (sizesData) {
    var names = [];
    for(var i = 0; i < sizesData.length; i++) {
      names.push(sizesData[i].name);
    }
    return names;
  },

  createDataSummaryObj: function (categoryName, subCategoryName, name, description, sizeNames, moduleClassName) {

    var obj = {};

    obj.categoryName = categoryName;
    obj.subCategoryName = subCategoryName;
    obj.name = name;
    obj.description = description;
    obj.cssClassName = _PropertyHelper.getCssClassName(name);
    obj.sizeNames = sizeNames.length > 0 ? sizeNames : 'None';
    obj.sizeCssClassNames = sizeNames.length > 0 ? _PropertyHelper.getSizeCssClassNames(name, sizeNames) : 'N/A';
    obj.moduleClassName = moduleClassName ? moduleClassName : 'None';

    return obj;

  },

  createDefinitionDataObj: function (categoryName, subCategoryName, name, description, sizeNames, moduleClassName) {

    var obj = {};

    // input
    obj.categoryName = categoryName;
    obj.subCategoryName = subCategoryName;
    obj.name = name;
    obj.description = description;

    // base
    obj.cssClassName = _PropertyHelper.getCssClassName(name);
    obj.sizeNames = sizeNames;
    obj.sizeCssClassNames = sizeNames.length > 0 ? _PropertyHelper.getSizeCssClassNames(name, sizeNames) : null;
    obj.fullCssClassName = _PropertyHelper.getFullCssClassName(name, sizeNames, null);
    obj.moduleClassName = moduleClassName;

    // sass template data
    obj.sassVarPrefix = _PropertyHelper.getSassVarPrefix(name);
    obj.sassTempColor = _config.components.templates.data.sass.tempColor;
    obj.sassTempBackground = _config.components.templates.data.sass.tempBackground;
    obj.sassTempPadding =_config.components.templates.data.sass.tempPadding;

    // docs
    obj.docsPageId = _DocsDataHelper.getDocsDataPageItemId(name);
    obj.dataModuleAttr = obj.moduleClassName ? ' data-module=\"' + obj.moduleClassName + '\"' : '';
    obj.exampleHtml = '<div class=\"' + obj.fullCssClassName + '\"' + obj.dataModuleAttr + '>\n\n\t\t\t<h4>' + name + '</h4>\n\n\t\t</div>';
    obj.codeHtml = '<div class=\"' + obj.fullCssClassName + '\"' + obj.dataModuleAttr + '>\n\n\t\t\t\t\t<h4>' + name + '</h4>\n\n\t\t\t\t</div>';

    // html
    var htmlPath = _config.components.paths.html;
    htmlPath += _PropertyHelper.getCategorySubCategoryNamePath(categoryName, subCategoryName, name);
    obj.htmlPath = htmlPath;
    obj.htmlFileName = _PropertyHelper.getHtmlDefinitionFileName(name);
    obj.htmlFilePath = obj.htmlPath + obj.htmlFileName;

    // notes
    var notesPath = _config.components.paths.notes;
    notesPath += _PropertyHelper.getCategorySubCategoryNamePath(categoryName, subCategoryName, name);
    obj.notesPath = notesPath;
    obj.notesFileName = _PropertyHelper.getNotesDefinitionFileName(name);
    obj.notesFilePath = obj.notesPath + obj.notesFileName;

    // sass
    var sassPath = _config.components.paths.sass;
    sassPath += _PropertyHelper.getCategorySubCategoryNamePath(categoryName, subCategoryName, name);
    obj.sassPath = sassPath;

    // sass definition
    obj.sassFileName = _PropertyHelper.getSassDefinitionFileName(name);
    obj.sassFilePath = obj.sassPath + obj.sassFileName;

    // sass sizes
    obj.sizes = _createSizeObjs(name, sizeNames, sassPath);

    // sass import
    obj.sassImportFileName = _PropertyHelper.getSassImportFileName(name);
    obj.sassImportFilePath = obj.sassPath + obj.sassImportFileName;
    obj.sassImports = '@import \'' + obj.sassFileName + '\';\n';
    for (var i = 0; i < obj.sizes.length; i++) {
      obj.sassImports += '@import \'' + obj.sizes[i].sassFileName + '\';\n';
    }

    // module
    obj.modulePath = _config.components.paths.modules;
    obj.moduleFileName = _PropertyHelper.getModuleFileName(name);
    obj.moduleFilePath = obj.modulePath + obj.moduleFileName;

    return obj;

  },

  createModifierDataObj: function (categoryName, subCategoryName, name, modifierName, modifierDescription) {

    var obj = {};

    // input
    obj.categoryName = categoryName;
    obj.subCategoryName = subCategoryName;
    obj.name = name;
    obj.modifierName = modifierName;
    obj.modifierDescription = modifierDescription;

    // base
    obj.cssClassName = _PropertyHelper.getModifierCssClassName(name, modifierName);
    obj.fullCssClassName = _PropertyHelper.getFullCssClassName(name, null, modifierName);

    // html
    var htmlPath = _config.components.paths.html;
    htmlPath += _PropertyHelper.getCategorySubCategoryNamePath(categoryName, subCategoryName, name);
    obj.htmlPath = htmlPath;
    obj.htmlFileName = _PropertyHelper.getHtmlModifierFileName(name, modifierName);
    obj.htmlFilePath = obj.htmlPath + obj.htmlFileName;

    // notes
    var notesPath = _config.components.paths.notes;
    notesPath += _PropertyHelper.getCategorySubCategoryNamePath(categoryName, subCategoryName, name);
    obj.notesPath = notesPath;
    obj.notesFileName = _PropertyHelper.getNotesModifierFileName(name, modifierName);
    obj.notesFilePath = obj.notesPath + obj.notesFileName;

    // sass
    var sassPath = _config.components.paths.sass;
    sassPath += _PropertyHelper.getCategorySubCategoryNamePath(categoryName, subCategoryName, name);
    obj.sassPath = sassPath;
    obj.sassFileName = _PropertyHelper.getSassModifierFileName(name, modifierName);
    obj.sassFilePath = obj.sassPath + obj.sassFileName;

    return obj;

  }

};
