// generator-jaywing-atomic/lib/components/FileManager.js
//-------------------------------------------------------
//  Library: Data Syncer
//-------------------------------------------------------

var _fs = require('fs');

var _config = require('../../config');
var _DocsDataHelper = require('../../lib/components/DocsDataHelper');

//-------------------------
//  Private
//-------------------------

var _generator;

function _getDocsData() {
  var path = _config.docs.files.global;
  var data = JSON.parse(_fs.readFileSync(path, 'utf8'));
  return data;
}

function _addSassDefinitionFile(data) {
  var path = data.sassFilePath;
  var template = 'definition.scss';
  _generator.fs.copyTpl(_generator.templatePath(template), _generator.destinationPath(path), data);
}

function _addSassSizeFiles(data) {
  for (var i = 0; i < data.sizes.length; i++) {
    var sizeData = data.sizes[i];
    var path = sizeData.sassFilePath;
    var template = 'size.scss';
    _generator.fs.copyTpl(_generator.templatePath(template), _generator.destinationPath(path), sizeData);
  }
}

function _addSassImportFile(data) {
  var path = data.sassImportFilePath;
  var template = 'imports.scss';
  _generator.fs.copyTpl(_generator.templatePath(template), _generator.destinationPath(path), data);
}

//-------------------------
//  Public
//-------------------------

var self = module.exports = {

  init: function (generator) {

    _generator = generator;

  },

  addDefinitionToDocsData: function (data) {

    var docsDataObj = _getDocsData();
    var pageObj = _DocsDataHelper.extractPageObj(docsDataObj, data.categoryName, data.subCategoryName);
    if(!pageObj) {
      pageObj = _DocsDataHelper.createPageObj(data.categoryName, data.subCategoryName);
      _DocsDataHelper.addPageObjToDocsDataObj(docsDataObj, pageObj);
    }
    var pageItemObj = _DocsDataHelper.createPageItemObj(data.name, data.categoryName);
    _DocsDataHelper.addPageItemObjToPageObj(pageObj, pageItemObj);

    // Write to docs data
    var path = _config.docs.files.global;
    var data = JSON.stringify(docsDataObj, null, 2);
    _fs.writeFile(path, data);

  },

  addDefinitionHtmlFiles: function (definitionData) {

    var data = definitionData;
    var path = data.htmlFilePath;
    var template = 'definition.html';
    _generator.fs.copyTpl(_generator.templatePath(template), _generator.destinationPath(path), data);

  },

  addDefinitionNotesFiles: function (definitionData) {

    var data = definitionData;
    var path = data.notesFilePath;
    var template = 'definition.md';
    _generator.fs.copyTpl(_generator.templatePath(template), _generator.destinationPath(path), data);

  },

  addDefinitionSassFiles: function (definitionData) {

    _addSassDefinitionFile(definitionData);
    _addSassSizeFiles(definitionData);
    _addSassImportFile(definitionData);

  },

  addDefinitionModuleFiles: function (definitionData) {

    var data = definitionData;
    var path = data.moduleFilePath;
    var template = 'definition.js';
    _generator.fs.copyTpl(_generator.templatePath(template), _generator.destinationPath(path), data);

  },

  addModifierHtmlFiles: function (modifierData) {

    var data = modifierData;
    var path = data.htmlFilePath;
    var template = 'modifier.html';
    _generator.fs.copyTpl(_generator.templatePath(template), _generator.destinationPath(path), data);

  },
  
  addModifierNotesFiles: function (modifierData) {

    var data = modifierData;
    var path = data.notesFilePath;
    var template = 'modifier.md';
    _generator.fs.copyTpl(_generator.templatePath(template), _generator.destinationPath(path), data);

  },
  
  addModifierSassFiles: function (modifierData) {

    var data = modifierData;
    var path = data.sassFilePath;
    var template = 'modifier.scss';
    _generator.fs.copyTpl(_generator.templatePath(template), _generator.destinationPath(path), data);

  }

};
