// generator-jaywing-atomic/lib/components/manager.js
//-------------------------------------------------------
//  Library: Data Manager
//-------------------------------------------------------

var _config = require('../../data/config');
var _miner = require('../../lib/components/miner');
var _helper = require('../../lib/components/helper');

//-------------------------
//  Private
//-------------------------

var _data = {};

//-------------------------
//  Public
//-------------------------

var self = module.exports = {

  init: function () {

    // init
    _data.componentData = {};

    // defaults
    _data.componentData.categories = _config.componentData.categories;
    _data.componentData.sizes = _config.componentData.sizes;

  },

  getCategoryNames: function () {
    var names = _helper.getCategoryNamesFromCategoriesData(_data.componentData.categories);
    return names;
  },

  getNames: function (category) {
    var names = _miner.getNamesFromSassFolders(category);
    return names;
  },

  getSizeNames: function () {
    var names = _helper.getSizeNamesFromSizesData(_data.componentData.sizes);
    return names;
  },

  getModuleNames: function () {
    var names = _miner.getModuleNamesFromModuleFiles();
    return names;
  },

  addComponent: function (categoryName, name, description, sizeNames, moduleName) {

  },

  deleteComponent: function (categoryName, name) {

  }

};
