// generator-jaywing-atomic/lib/components/manager.js
//-------------------------------------------------------
//  Library: Data Manager
//-------------------------------------------------------

var _config = require('../../data/config');
var _miner = require('../../lib/components/miner');
var _helper = require('../../lib/components/helper');

//-------------------------
//  Private
//-------------------------

var _data = {};

//-------------------------
//  Public
//-------------------------

var self = module.exports = {

  init: function () {

    // init
    _data.componentData = {};

    // defaults
    _data.componentData.categories = _config.componentData.categories;
    _data.componentData.sizes = _config.componentData.sizes;

  },

  getCategoryNames: function () {
    var names = _helper.getDataCategoryNamesFromCategories(_data.componentData.categories);
    return names;
  },

  getSubCategoryNames: function (categoryName) {

    var names = [];

    // Find the corresponding category in the data
    if(categoryName == 'Core'){
      var categories = _data.componentData.categories;
      for (var i = 0; i < categories.length; i++) {
        if(categories[i].name == categoryName) {
          var category = categories[i];
          names = _helper.getDataSubCategoryNamesFromCategory(category);
          break;
        }
      }
    } else {
      names = _miner.getSubCategoryNamesFromSassFolders(categoryName);
    }

    return names;

  },

  getNames: function (category) {
    var names = _miner.getNamesFromSassFolders(category);
    return names;
  },

  getSizeNames: function () {
    var names = _helper.getDataSizeNamesFromSizes(_data.componentData.sizes);
    return names;
  },

  getModuleClassNames: function () {
    var names = _miner.getModuleClassNamesFromModuleFiles();
    return names;
  },

  addComponent: function (categoryName, name, description, sizeNames, moduleName) {

  },

  deleteComponent: function (categoryName, name) {

  }

};
