// generator-jaywing-atomic/lib/components/PropertyHelper.js
//-------------------------------------------------------
//  Library: Data Helper
//-------------------------------------------------------

var _ = require('lodash');

var _config = require('../../config');

//-------------------------
//  Private
//-------------------------

function _getPathString(string) {
  return string.replace(/\s+/g, '-').toLowerCase();
}

//-------------------------
//  Public
//-------------------------

var self = module.exports = {

  getCategoryPath: function (categoryName) {
    var pathStr = _getPathString(categoryName);
    return pathStr + '/';
  },

  getSubCategoryPath: function (subCategoryName) {
    var pathStr = _getPathString(subCategoryName);
    return pathStr + '/';
  },

  getNamePath: function (name) {
    var pathStr = _getPathString(name);
    return pathStr + '/';
  },

  getFormattedName: function (inputName) {
    return _.startCase(_.toLower(inputName));
  },

  getFormattedSubCategoryName: function (inputName) {
    return _.startCase(_.toLower(inputName));
  },

  getCssClassName: function (name) {
    return name.replace(/\s+/g, '');
  },

  getSassVarPrefix: function (name) {
    return name.replace(/\s+/g, '-').toLowerCase();
  },

  getSizeSassIncludeName: function (sizeName) {
    return sizeName.toLowerCase();
  },

  getSizeCssClassName: function (name, sizeName) {
    var name = this.getCssClassName(name);
    var result = name + sizeName;
    return result;
  },

  getSizeCssClassNames: function (name, sizeNames) {
    var results = [];
    for (var i = 0; i < sizeNames.length; i++) {
      var name = this.getCssClassName(name);
      var cssClassName = this.getSizeCssClassName(name, sizeNames[i]);
      results.push(cssClassName);
    }
    return results;
  },

  getModifierCssClassSuffix: function (modifierName) {
    var strs = modifierName.toLowerCase().match(/[^\s-]+/g);
    var result = strs[0].toLowerCase();
    if(strs.length > 1){
      for(var i = 1; i < strs.length; i++) {
        result += _.upperFirst(strs[i]);
      }
    }
    return result;
  },

  getModifierCssClassName: function (name, modifierName) {
    var css = this.getCssClassName(name);
    var suffix = this.getModifierCssClassSuffix(modifierName);
    return css + '--' + suffix;
  },

  getModifierCssClassNames: function (name, modifierNames) {
    var results = [];
    for (var i = 0; i < modifierNames.length; i++) {
      var modifier = this.getModifierCssClassName(name, modifierNames[i]);
      results.push(modifier);
    }
    return results;
  },

  getFullCssClassName: function (name, sizeNames, modifierNames) {
    var css = this.getCssClassName(name);
    var sizes = this.getSizeCssClassNames(name, sizeNames);
    var modifiers = this.getModifierCssClassNames(name, modifierNames);
    var result = css;
    for (var i = 0; i < sizes.length; i++) {
      result += ' ' + sizes[i];
    }
    for (var i = 0; i < modifiers.length; i++) {
      result += ' ' + modifiers[i];
    }
    return result;
  },

  getModuleClassName: function (name) {
    return name.replace(/\s+/g, '');
  },

  getSassDefinitionFileName: function (name) {
    var path = _getPathString(name);
    return 'component.' + path + '.definition.scss';
  },

  getSassSizeFileName: function (name, sizeName) {
    var result = 'component.';
    result += _getPathString(name);
    result += '.';
    result += _getPathString(sizeName);
    result += '.definition.scss';
    return result;
  },

  getSassImportFileName: function (name) {
    var result = 'component.';
    result += _getPathString(name);
    result += '.scss';
    return result;
  }

};
